antelope_contracts/contracts/erc20/src/erc20.cpp

```diff
@@ -8,84 +8,247 @@
+        }
+    };
+
+    pack_uint32(constructor_data, (uint8_t)erc20_precision); // offset 0
+    pack_uint32(constructor_data, 128);                      // offset 32
+    pack_uint32(constructor_data, 192);                      // offset 64
+    pack_uint32(constructor_data, 256);                      // offset 96
+    pack_string(constructor_data, evm_token_name);           // offset 128
+    pack_string(constructor_data, evm_token_symbol);         // offset 192
+    pack_string(constructor_data, eos_contract_name.to_string()); // offset 256
+
+    pack_uint32(call_data, 64);
+    pack_string(call_data, constructor_data);
 
-    memcpy(new_config.erc20_addr, deploy_addr.bytes, kAddressLength);
+    bytes to = {};
+    bytes value_zero; 
+    value_zero.resize(32, 0);
+
+     // required account opened in evm_runtime
+    call_action call_act(evm_account, {{get_self(), "active"_n}});
+    call_act.send(get_self(), to, value_zero, call_data, evm_init_gaslimit);
+
+    evmc::address proxy_contract_addr = silkworm::create_address(reserved_addr, nonce); 
 
-    _config.set(new_config, get_self());
+    token_table.emplace(_self, [&](auto &v) {
+        v.eos_contract_name = eos_contract_name;
+        v.address.resize(kAddressLength, 0);
+        memcpy(&(v.address[0]), proxy_contract_addr.bytes, kAddressLength);
+        v.min_deposit = min_deposit;
+        v.deposit_fee = deposit_fee;
+        v.erc20_precision = erc20_precision;
+    });
 }
 
-void erc20::onbridgemsg(name receiver, const bytes& sender, const time_point& timestamp, const bytes& value, const bytes& data) {
-    require_auth(evm_account);
+void erc20::onbridgemsg(const bridge_message_t &message) {
+
+    check(get_sender() == evm_account, "invalid sender of onbridgemsg");
+
+    const bridge_message_v0 &msg = std::get<bridge_message_v0>(message);
+    check(msg.receiver == _self, "invalid message receiver");
+
+    checksum256 addr_key = make_key(msg.sender);
+
+    token_table_t token_table(_self, _self.value);
+    auto index = token_table.get_index<"by.address"_n>();
+    auto itr = index.find(addr_key);
 
-    // TODO: this API will change
+    check(itr != index.end() && itr->address == msg.sender, "ERC-20 token not registerred");
 
-    check(data.size() == kAddressLength + 32 , "invalid data size");
-    // TODO: verify from, decode data
-    intx::uint256 amount = intx::be::unsafe::load<intx::uint256>((const uint8_t*)data.data() + kAddressLength);
-    evmc::address dest = {};
-    memcpy(dest.bytes, data.data(), kAddressLength);
-    auto to = silkworm::extract_reserved_address(dest);
+    check(msg.data.size() >= 4, "not enough data in bridge_message_v0");
+    if (*(uint32_t *)&(msg.data[0]) == 0xe5323365 /* big-endian 0x653332e5 */) {
+        //sha("bridgeTransferV0(address,uint,string)") = 0xe5323365
+        check(msg.data.size() >= 4 + kAddressLength + 32, "not enough data in bridge_message_v0 of application type 0xe5323365");
```

Comment:
seems not in this case, because in the contract.sol it is using abi.encodePacked which results in compacked encoded binary
```
bytes memory n_bytes = abi.encodePacked(app_version, to, amount, memo);
```